otherlibs/systhreads/st_stubs.c

```diff
@@ -410,18 +411,20 @@ static void caml_thread_domain_stop_hook(void) {
   };
 }
 
-static void caml_thread_domain_initialize_hook(void)
+/* Returns 0 if initialization failed */
```

Comment:
This semantics is the opposite of the semantics of other thread operations like `pthread_mutex_init` and `pthread_cond_init` where `0` is success and non-zero is an error. Wondered whether there was a reason to do it this way?

Reply:
Agreed with @kayceesrk.   It's the opposite of the convention established in `st_masterlock_init` in this diff.

Reply:
The reasoning was that the other ones are internal, while this one is exported in some sense (my view was probably influenced by hacking heavily the runtime hooks in the past).

The new version raises an exception instead. This is mostly useful when initializing threads on the first domain, which was already prepared to receive an exception.

When fixing domain_thread_func to handle errors, this will need to be changed to returning an encoded exception.

Reply:
Since I feel that my comment was not very clear: I am backtracking a bit on returning an error code and I keep the exception-raising behaviour of caml_stat_alloc. I add a check to raise an exception on faulty mutex initialization. This lets me improve the behaviour when initializing threads on the first domain, without fixing or making worse the behaviour on initialization for other domains. This last fix will have to come in another PR (as mentioned already).